technical-reference/reference-technical-articles/diamond-comptroller.md

```diff
@@ -0,0 +1,83 @@
+
+**Step 2:** Division of the comptroller into multiple facets based on the separation of concerns.
+
+Following the same eternal storage pattern i.e. inheriting storage contract for storage slots. This basically replaced the AppStorage of Diamond pattern.
+
+Comptroller storage contains the facet addresses, mapping of the function selectors to the facet address and mapping of all selectors to the facet address.
+
+## User interaction with new diamond pattern
+The user interaction remains consistent, wherein users will continue to interact with the Unitroller proxy contract. However, a significant modification has occurred in the way the implementation contract is invoked. Now, the Diamond proxy serves as the implementation contract for the Unitroller, and the Comptroller's implementation has been subdivided into multiple facets. These facets will function as the various components of the Diamond proxy.
+
+Following a user's interaction with the Unitroller, the `delegateCall` operation will be directed towards the Diamond proxy. Once the `delegateCall` reaches the Diamond proxy, it will analyze the function selector and determine the appropriate facet address to which the `delegateCall` will be forwarded.
+
+## How will new Upgrades work?
+**Scenario:** If a new state variable is to be introduced for an upgrade, which is being used by all or few facets.
+
+Example: Add a new state variable in the `comptrollerStorage` by extending the storage to `ComptrollerV14Storage` just like the new states are added for the facets by extending the previous storage.
+
+**Scenario:** Add a new facet to the Diamond or update the existing facet by add/replace/remove function selectors.
+
+Deploy the facet with the functions/methods that need to be added to the Comptroller Implementation.
+
+Execute the `diamondCut` method(admin access) to add the new facet to the `comptrollerStorage`.
+
+To execute the `diamondCut` through VIP to add a new facet or add/replace/remove functions in the existing facet, there will be a need to pass the cut as an argument which can be generated by the facet-cut-params-generator script through the `generateCutParams` method by making changes to the constants used in the script according the actions need to perform on provided function selectors.
+
+## Diamond proxy implementation for Comptroller in the Core pool
+The Comptroller of the core pool is divided into 4 facets and 2 parent contracts extended by the facets. Facets will receive the function call through a chained delegateCall from Unitroller, and each facet holds its own responsibility as explained below:
+
+### Facets
+**PolicyFacet:** This facet contains all the external pre-hook functions related to vToken.
```

Comment:
Include links to natspec for facets and parent contracts

Reply:
Should i add github links for their netspec files?

Reply:
I would link to them in gitbook. The links can be relative to their directory in this repo

Reply:
Done [aa851b5](https://github.com/VenusProtocol/venus-protocol-documentation/pull/69/commits/aa851b5af5063542068ad5e6bf0ab9fb41087b00)